nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java

```diff
@@ -95,37 +90,32 @@ public void onMatchMethod(
         return;
       }
 
-      supportedMethods.add(state.getPath());
-    }
-  }
-
-  @Override
-  public void onMatchReturn(NullAway analysis, ReturnTree tree, VisitorState state) {
-    final TreePath enclosingMethod =
-        NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(state.getPath());
-    if (!supportedMethods.contains(enclosingMethod)) {
-      return;
-    }
-
-    final Nullness nullness =
-        analysis
-            .getNullnessAnalysis(state)
-            .getNullnessForContractDataflow(
-                new TreePath(state.getPath(), tree.getExpression()), state.context);
-
-    if (nullness == Nullness.NULLABLE || nullness == Nullness.NULL) {
-      reportMatchForContractIssue(
-          tree,
-          "@Contract might not be followed, as returning @Nullable from method with @NonNull return expected in contract",
-          analysis,
-          state);
+      // we scan the method tree for the return nodes and check the contract
+      new TreePathScanner<Void, Void>() {
+        @Override
+        public Void visitReturn(ReturnTree returnTree, Void unused) {
+
+          final VisitorState returnState =
+              state.withPath(TreePath.getPath(state.getPath(), returnTree));
```

Comment:
Wonder how expensive this search is or even the traversal itself. It might be more expensive than traversing up the AST to find the enclosing method, actually, but we can leave the optimization for later, as this is behind an option and we can always profile with/without this handler.

Reply:
I checked and the dataflow analysis result is getting cached (at least based on my reading of the code).  This is an extra traversal, but the previous approach was traversing up the AST to find the enclosing method for all `return` statements in the program (to check if the method was one with a `@Contract` annotation).  This should only do work for methods where we've already found a relevant `@Contract` annotation I think.

Reply:
Right. So the initial one was an approximately `O(M * log N)` cost where `M` is the number of all methods in the analyzed code and `N` the number of nodes in the average method AST (handwaving a lot of things about tree's being balanced and amortizing across method ASTs of vastly different sizes), while this is `O(C * N)` where `C` is the number of methods _with contract annotations_. The key is that we expect `C << M` to an even greater degree than `log N << N`. Correct?

Reply:
Yes, since method sizes tend to be small, that is my intuition 🙂.  Plus, the previous approach required maintaining some extra instance state to cache which methods were `@Contract` methods, and that set had to be cleared on each top-level class to avoid a leak.  Probably best to avoid such things unless they are really worth it